Factor
- A factor is a number that divides evenly into another number.  For example, 2 is a factor of 4 because 2 divides evenly into 4.  3 and 4 are both factors of 12 because they both divide evenly into 12.

In [9]:
def getFactors(n):
    result = []
    for i in range (2, n+1):
        if n % i == 0:
            result.append(i)
    return result

num = int(input("Input a number: "))
print("The factors of", num, "is", getFactors(num))

The factors of 24 is [2, 3, 4, 6, 8, 12, 24]


Prime number
- A number with exactly 2 factors -- 1 and itself.  For example, 2 is a prime number because its only 2 factors are 1 and 2.  41 is also a prime number because it can only be divided by 1 and 41.

In [25]:
def isPrime(n):
    for i in range(2, int(n**0.5) + 1):
        if n % i == 0:
            return False
    return True

# Driver program
print("Below are all prime numbers till 10000")
for n in range (2, 10000):
    if isPrime(n):
        print(n, " is a prime number")

Below are all prime numbers till 10000
2  is a prime number
3  is a prime number
5  is a prime number
7  is a prime number
11  is a prime number
13  is a prime number
17  is a prime number
19  is a prime number
23  is a prime number
29  is a prime number
31  is a prime number
37  is a prime number
41  is a prime number
43  is a prime number
47  is a prime number
53  is a prime number
59  is a prime number
61  is a prime number
67  is a prime number
71  is a prime number
73  is a prime number
79  is a prime number
83  is a prime number
89  is a prime number
97  is a prime number
101  is a prime number
103  is a prime number
107  is a prime number
109  is a prime number
113  is a prime number
127  is a prime number
131  is a prime number
137  is a prime number
139  is a prime number
149  is a prime number
151  is a prime number
157  is a prime number
163  is a prime number
167  is a prime number
173  is a prime number
179  is a prime number
181  is a prime number
191  is a prime number
1

Composite number
- A number with more than 2 factors.  For example, 9 is a composite number because its factors are 1, 3, and 9.  Please note that number 1 is neither prime nor composite.  

In [ ]:
# Check if a number is composite.
def isComposite(n):
    # Corner cases
    if (n <= 1):
        return False
    if (n <= 3):
        return False
    # This is checked so that we can skip
    # middle five numbers in below loop
    if (n % 2 == 0 or n % 3 == 0):
        return True
    i = 5
    while(i * i <= n):
        if (n % i == 0 or n % (i + 2) == 0):
            return True
        i = i + 6
    return False

# Driver program
print("Below are all composite numbers till 10000")
n = 2
for n in range(10000):
    if isComposite(n):
        print(n , " is a composite number")

Perfect number
- A number is said to be perfect if it is equal to the sum of all its factors (for obvious reasons the list of factors being considered does not include the number itself).  6 = 3 + 2 + 1, hence 6 is perfect.  28 is another example since 1 + 2 + 4 + 7 + 14 is 28.  Please note that the number 1 is not a perfect number.

In [24]:
# check if a number is perfect or not
# Returns true if n is perfect

def isPerfect(n):
    # To store sum of divisors
    sum = 1
    # Find all divisors and add them
    i = 2
    while i * i <= n:
        if n % i == 0:
            sum = sum + i + n/i
        i += 1
    # If sum of divisors is equal to n, then n is a perfect number
    return (True if sum == n and n!=1 else False)
 
# Driver program
print("Below are all perfect numbers till 10000")
n = 2
for n in range (10000):
    if isPerfect (n):
        print(n , " is a perfect number")

Below are all perfect numbers till 10000
6  is a perfect number
28  is a perfect number
496  is a perfect number
8128  is a perfect number


Abundant number
- A number is considered to be abundant if the sum of its factors (aside from the number) is greater than the number itself.  For example, 12 is abundant since 1 + 2 + 3 + 4 + 6 = 16 > 12.  However, a number like 15, where the sum of the factors is 1 + 3 + 5 = 9, is not abundant.

In [23]:
def isAbundant(n):  
    fctr_sum = sum([fctr for fctr in range(1, n) if n % fctr == 0])  
    return fctr_sum > n  

# Driver program
print("Below are all abundant numbers till 10000")
n = 2
for n in range (10000):
    if isAbundant(n):
        print(n , " is an abundant number")

Below are all abundant numbers till 10000
12  is an abundant number
18  is an abundant number
20  is an abundant number
24  is an abundant number
30  is an abundant number
36  is an abundant number
40  is an abundant number
42  is an abundant number
48  is an abundant number
54  is an abundant number
56  is an abundant number
60  is an abundant number
66  is an abundant number
70  is an abundant number
72  is an abundant number
78  is an abundant number
80  is an abundant number
84  is an abundant number
88  is an abundant number
90  is an abundant number
96  is an abundant number
100  is an abundant number
102  is an abundant number
104  is an abundant number
108  is an abundant number
112  is an abundant number
114  is an abundant number
120  is an abundant number
126  is an abundant number
132  is an abundant number
138  is an abundant number
140  is an abundant number
144  is an abundant number
150  is an abundant number
156  is an abundant number
160  is an abundant number
162  is

Triangular number
- The triangular number Tn is a number that can be represented in the form of a triangular grid of points where the first row contains a single element and each subsequent row contains one more element than the previous one (see figure below).  
- For example, 3 is the 2nd triangular number, since 2 (3) / 2 = 3.  As another example, 15 is the 5th triangular number, since 5 (6) / 2 = 15.  However, 14 is not triangular, since there is no x such that x (x + 1) / 2 = n.

In [22]:
# Check if a number is a triangular number using simple approach.
# Returns True if 'num' is triangular, else False

def isTriangular(num):
    # Base case
    if (num < 0):
        return False
    # A Triangular number must be
    # sum of first n natural numbers
    sum, n = 0, 1
    while(sum <= num):
        sum = sum + n
        if (sum == num):
            return True
        n += 1
    return False
# Driver program
print("Below are all triangular numbers till 10000")
n = 2
for n in range (10000):
    if isTriangular(n):
        print(n , " is a triangular number")

Below are all triangular numbers till 10000
1  is a triangular number
3  is a triangular number
6  is a triangular number
10  is a triangular number
15  is a triangular number
21  is a triangular number
28  is a triangular number
36  is a triangular number
45  is a triangular number
55  is a triangular number
66  is a triangular number
78  is a triangular number
91  is a triangular number
105  is a triangular number
120  is a triangular number
136  is a triangular number
153  is a triangular number
171  is a triangular number
190  is a triangular number
210  is a triangular number
231  is a triangular number
253  is a triangular number
276  is a triangular number
300  is a triangular number
325  is a triangular number
351  is a triangular number
378  is a triangular number
406  is a triangular number
435  is a triangular number
465  is a triangular number
496  is a triangular number
528  is a triangular number
561  is a triangular number
595  is a triangular number
630  is a triangular

Narcissistic number
- A positive integer is called narcissistic if it is equal to the sum of its own digits each raised to the power of the number of digits.  For example, 153 is narcissistic because 1^3 + 5^3 + 3^3 = 1 + 125 + 27 = 153.  Note that by this definition all single digit numbers are narcissistic.

In [21]:
# Returns true if n is Narcissistic number
def isNarcissistic(x):
    n = x
    digits = 1
    while (n>=10):
        n = n // 10
        digits += 1

    dup = x; sm = 0
    # Calculates the sum of digits
    # raised to power
    while (dup) :
        sm = sm + pow(dup % 10, digits)
        dup = dup // 10
    return (x == sm)

# Driver program
print("Below are all Narcissistic numbers till 1000")
for n in range(1, 10000):
    if isNarcissistic(n):
        print(n , " is a Narcissistic number")

Below are all Narcissistic numbers till 1000
1  is a Narcissistic number
2  is a Narcissistic number
3  is a Narcissistic number
4  is a Narcissistic number
5  is a Narcissistic number
6  is a Narcissistic number
7  is a Narcissistic number
8  is a Narcissistic number
9  is a Narcissistic number
153  is a Narcissistic number
370  is a Narcissistic number
371  is a Narcissistic number
407  is a Narcissistic number
1634  is a Narcissistic number
8208  is a Narcissistic number
9474  is a Narcissistic number
